File: skills/grain/forge/src/fanin.nim
# fanin.nim -- module-level symbol fan-in counter.
#
# Transliteration of forge/src/fanin.py. Byte-identical output is the
# requirement, not merely equivalent output (shared/forge.md §D). Read
# fanin.py first; this file follows its structure function for function.
#
# Stdlib only: std/json, std/algorithm, std/tables, std/strutils.

import std/json
import std/algorithm
import std/tables
import std/strutils

type
  ContractError = object of ValueError
    ## Raised when stdin does not satisfy the fanin contract.

  Tag = object
    kind: string
    name: string
    path: string
    line: int

  Request = object
    files: seq[string]
    tags: seq[Tag]

  Symbol = object
    name: string
    path: string
    line: int
    internal: int
    external: int

const TagKinds = ["call", "def"] ## sorted; mirrors sorted(TAG_KINDS) in fanin.py

# --- the parse boundary --------------------------------------------------
# Everything in this section is the one place a malformed document is
# allowed to surface as a raised error rather than being coerced.

proc isValidUtf8(s: string): bool =
  ## Hand-rolled UTF-8 validator (no std/unicode — outside the allowed
  ## import list). Rejects overlong encodings, lone surrogates, and
  ## codepoints above U+10FFFF, same as CPython's strict UTF-8 codec.
  var i = 0
  let n = s.len
  while i < n:
    let b0 = byte(ord(s[i]))
    if b0 < 0x80'u8:
      inc i
    elif (b0 and 0xE0'u8) == 0xC0'u8:
      if i + 1 >= n or b0 < 0xC2'u8:
        return false
      let b1 = byte(ord(s[i + 1]))
      if (b1 and 0xC0'u8) != 0x80'u8:
        return false
      i += 2
    elif (b0 and 0xF0'u8) == 0xE0'u8:
      if i + 2 >= n:
        return false
      let b1 = byte(ord(s[i + 1]))
      let b2 = byte(ord(s[i + 2]))
      if (b1 and 0xC0'u8) != 0x80'u8 or (b2 and 0xC0'u8) != 0x80'u8:
        return false
      if b0 == 0xE0'u8 and b1 < 0xA0'u8:
        return false # overlong
      if b0 == 0xED'u8 and b1 >= 0xA0'u8:
        return false # UTF-16 surrogate range, invalid in UTF-8
      i += 3
    elif (b0 and 0xF8'u8) == 0xF0'u8:
      if i + 3 >= n:
        return false
      let b1 = byte(ord(s[i + 1]))
      let b2 = byte(ord(s[i + 2]))
      let b3 = byte(ord(s[i + 3]))
      if (b1 and 0xC0'u8) != 0x80'u8 or (b2 and 0xC0'u8) != 0x80'u8 or
          (b3 and 0xC0'u8) != 0x80'u8:
        return false
      if b0 == 0xF0'u8 and b1 < 0x90'u8:
        return false # overlong
      if b0 > 0xF4'u8 or (b0 == 0xF4'u8 and b1 >= 0x90'u8):
        return false # beyond U+10FFFF
      i += 4
    else:
      return false
  true

proc ensureUtf8(value: string, field: string) =
  if not isValidUtf8(value):
    raise newException(ContractError, field & " is not valid UTF-8")

proc requireStr(node: JsonNode, field: string): string =
  if node.isNil or node.kind != JString:
    raise newException(ContractError, field & " must be a string")
  result = node.getStr()
  ensureUtf8(result, field)

proc requireInt(node: JsonNode, field: string): int =
  if node.isNil or node.kind != JInt:
    raise newException(ContractError, field & " must be an integer")
  node.getInt()

proc getField(obj: JsonNode, key: string, context: string): JsonNode =
  if obj.isNil or obj.kind != JObject:
    raise newException(ContractError, context & " must be a JSON object")
  if not obj.hasKey(key):
    return nil
  obj[key]

proc parseRequest(raw: JsonNode): Request =
  ## Walk parsed JSON, validate every field, return plain records.
  ## Raises ContractError on anything unexpected rather than coercing it.
  if raw.isNil or raw.kind != JObject:
    raise newException(ContractError, "top level must be a JSON object")

  let filesNode = getField(raw, "files", "top level")
  if filesNode.isNil or filesNode.kind != JArray:
    raise newException(ContractError, "'files' must be a JSON array")
  var files: seq[string] = @[]
  for f in filesNode.elems:
    files.add(requireStr(f, "files[]"))

  let tagsNode = getField(raw, "tags", "top level")
  if tagsNode.isNil or tagsNode.kind != JArray:
    raise newException(ContractError, "'tags' must be a JSON array")

  var tags: seq[Tag] = @[]
  for i, entry in tagsNode.elems:
    let ctx = "tags[" & $i & "]"
    if entry.isNil or entry.kind != JObject:
      raise newException(ContractError, ctx & " must be a JSON object")
    let kind = requireStr(getField(entry, "kind", ctx), ctx & ".kind")
    if kind notin TagKinds:
      raise newException(
        ContractError, ctx & ".kind must be one of [call, def], got '" & kind & "'"
      )
    let name = requireStr(getField(entry, "name", ctx), ctx & ".name")
    let path = requireStr(getField(entry, "path", ctx), ctx & ".path")
    let line = requireInt(getField(entry, "line", ctx), ctx & ".line")
    tags.add(Tag(kind: kind, name: name, path: path, line: line))

  Request(files: files, tags: tags)

# --- computation -----------------------------------------------------------
# Pure arithmetic. tags[] is survey's already-merged view of a scope: ctags
# definitions (kind "def") and grep-derived call sites (kind "call"), keyed
# on name. A call is never disambiguated against imports — two definitions
# sharing a name in different files both receive credit for an external
# call under that name. That is a deliberate limit of a name-only heuristic
# (forge.md §E: it computes, grain concludes), not a bug.

proc computeSymbols(req: Request): seq[Symbol] =
  var scope = initTable[string, bool]()
  for f in req.files:
    scope[f] = true

  # Keys, not a table of Symbol: dedupes an exact-duplicate def without
  # losing two distinct definitions that merely share a name. defOrder
  # is first-seen order; the final sort below is what makes the output
  # deterministic, not this order.
  var seen = initTable[(string, string, int), bool]()
  var defOrder: seq[(string, string, int)] = @[]
  for tag in req.tags:
    if tag.kind == "def" and scope.hasKey(tag.path):
      let key = (tag.path, tag.name, tag.line)
      if not seen.hasKey(key):
        seen[key] = true
        defOrder.add(key)

  var callsByName = initTable[string, seq[string]]()
  for tag in req.tags:
    if tag.kind == "call":
      callsByName.mgetOrPut(tag.name, @[]).add(tag.path)

  var symbols: seq[Symbol] = @[]
  for key in defOrder:
    let (path, name, line) = key
    var internal = 0
    var external = 0
    if callsByName.hasKey(name):
      for p in callsByName[name]:
        if p == path:
          inc internal
        else:
          inc external
    symbols.add(
      Symbol(name: name, path: path, line: line, internal: internal, external: external)
    )

  # Deliberate re-sort, never a reliance on table iteration order:
  # (path, name) per the contract, with line as a deterministic tiebreak
  # for two definitions that share both. Nim's string `<` is byte order,
  # matching Python's codepoint order for valid UTF-8 (forge.md §D).
  symbols.sort proc(a, b: Symbol): int =
    if a.path != b.path:
      return cmp(a.path, b.path)
    if a.name != b.name:
      return cmp(a.name, b.name)
    cmp(a.line, b.line)

  symbols

# --- emission ----------------------------------------------------------
# Hand-written, not JsonNode's default `$`: it must match Python's
# json.dumps(sort_keys=True, separators=(",", ":"), ensure_ascii=True)
# byte for byte, which the default serializer does not (forge.md §D).

proc appendEscaped(buf: var string, s: string) =
  buf.add('"')
  var i = 0
  let n = s.len
  while i < n:
    let b0 = byte(ord(s[i]))
    if b0 < 0x80'u8:
      let c = char(b0)
      case c
      of '"':
        buf.add("\\\"")
      of '\\':
        buf.add("\\\\")
      of '\b':
        buf.add("\\b")
      of '\f':
        buf.add("\\f")
      of '\n':
        buf.add("\\n")
      of '\r':
        buf.add("\\r")
      of '\t':
        buf.add("\\t")
      else:
        if b0 < 0x20'u8 or b0 == 0x7F'u8:
          buf.add("\\u")
          buf.add(toLowerAscii(toHex(int(b0), 4)))
        else:
          buf.add(c)
      inc i
    else:
      # Multi-byte UTF-8 sequence -> decode one codepoint. Already
      # validated at the parse boundary, so the shape is trusted here.
      var cp: int
      var seqLen: int
      if (b0 and 0xE0'u8) == 0xC0'u8:
        cp = int(b0 and 0x1F'u8)
        seqLen = 2
      elif (b0 and 0xF0'u8) == 0xE0'u8:
        cp = int(b0 and 0x0F'u8)
        seqLen = 3
      else:
        cp = int(b0 and 0x07'u8)
        seqLen = 4
      for k in 1 ..< seqLen:
        let bk = byte(ord(s[i + k]))
        cp = (cp shl 6) or int(bk and 0x3F'u8)
      i += seqLen
      if cp > 0xFFFF:
        let cpp = cp - 0x10000
        let hi = 0xD800 + (cpp shr 10)
        let lo = 0xDC00 + (cpp and 0x3FF)
        buf.add("\\u")
        buf.add(toLowerAscii(toHex(hi, 4)))
        buf.add("\\u")
        buf.add(toLowerAscii(toHex(lo, 4)))
      else:
        buf.add("\\u")
        buf.add(toLowerAscii(toHex(cp, 4)))
  buf.add('"')

proc render(symbols: seq[Symbol]): string =
  var buf = newStringOfCap(64 + symbols.len * 48)
  buf.add("{\"symbols\":[")
  for idx, s in symbols:
    if idx > 0:
      buf.add(",")
    buf.add("{\"external\":")
    buf.add($s.external)
    buf.add(",\"internal\":")
    buf.add($s.internal)
    buf.add(",\"line\":")
    buf.add($s.line)
    buf.add(",\"name\":")
    appendEscaped(buf, s.name)
    buf.add(",\"path\":")
    appendEscaped(buf, s.path)
    buf.add("}")
  buf.add("]}")
  buf

# --- entry point ---------------------------------------------------------

proc main(): int =
  let rawText = stdin.readAll()
  if not isValidUtf8(rawText):
    stderr.writeLine("fanin: stdin is not valid UTF-8")
    return 1

  var parsed: JsonNode
  try:
    parsed = parseJson(rawText)
  except JsonParsingError as exc:
    stderr.writeLine("fanin: invalid JSON on stdin: " & exc.msg)
    return 1

  var request: Request
  try:
    request = parseRequest(parsed)
  except ContractError as exc:
    stderr.writeLine("fanin: " & exc.msg)
    return 1

  let symbols = computeSymbols(request)
  stdout.write(render(symbols))
  0

quit(main())

File: skills/grain/forge/src/fanin.py
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""fanin — module-level symbol fan-in counter.

Delegable target per ``shared/forge.md`` §E: pure arithmetic over
normalised occurrence records. It never parses source syntax and never
decides what a finding means — it counts.

``tags[]`` on stdin is ``survey``'s already-merged view of a scope: ctags
definitions (``kind: "def"``) and grep-derived call sites
(``kind: "call"``), keyed on the same ``name``. This script never
disambiguates a call against imports — two definitions sharing a name in
different files both receive credit for an external call under that name.
That is a known, deliberate limit of a name-only heuristic (see
``shared/forge.md`` §E: it computes, ``grain`` concludes), not a bug.

Contract (``shared/forge.md`` §D, §G):

* stdin  -- one JSON object: ``{"files": [str], "tags": [tag]}``
* stdout -- one JSON object, nothing else, no trailing newline
* stderr -- diagnostics, nothing else
* exit   -- 0 on success, non-zero on refusal. Never partial output.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from typing import Any, Final

TAG_KINDS: Final[frozenset[str]] = frozenset({"def", "call"})


class ContractError(ValueError):
    """Raised when stdin does not satisfy the fanin contract."""


@dataclass(frozen=True, slots=True)
class Tag:
    """One normalised occurrence: a definition or a call site."""

    kind: str
    name: str
    path: str
    line: int


@dataclass(frozen=True, slots=True)
class Request:
    """A validated, parsed stdin payload."""

    files: tuple[str, ...]
    tags: tuple[Tag, ...]


@dataclass(frozen=True, slots=True)
class Symbol:
    """One module-level symbol and its fan-in split."""

    name: str
    path: str
    line: int
    internal: int
    external: int


def _ensure_utf8(value: str, field: str) -> str:
    """Reject a string that cannot round-trip through UTF-8.

    The parse boundary is the one place this check belongs (``forge.md``
    §D) — everything downstream trusts that every string it holds is
    valid UTF-8.
    """
    try:
        value.encode("utf-8", errors="strict")
    except UnicodeEncodeError as exc:
        msg = f"{field} is not valid UTF-8: {value!r}"
        raise ContractError(msg) from exc
    return value


def _require_str(obj: Any, field: str) -> str:  # noqa: ANN401 -- parse boundary
    if not isinstance(obj, str):
        msg = f"{field} must be a string, got {type(obj).__name__}"
        raise ContractError(msg)
    return _ensure_utf8(obj, field)


def _require_int(obj: Any, field: str) -> int:  # noqa: ANN401 -- parse boundary
    if not isinstance(obj, int) or isinstance(obj, bool):
        msg = f"{field} must be an integer, got {type(obj).__name__}"
        raise ContractError(msg)
    return obj


def parse_request(raw: Any) -> Request:  # noqa: ANN401 -- the one place Any may appear
    """Walk parsed JSON, validate every field, return frozen dataclasses.

    This is the only function in the module where ``Any`` may appear and
    the only place a runtime type error can originate. It raises on
    anything unexpected rather than coercing it.
    """
    if not isinstance(raw, dict):
        msg = "top level must be a JSON object"
        raise ContractError(msg)

    files_raw = raw.get("files")
    if not isinstance(files_raw, list):
        msg = "'files' must be a JSON array"
        raise ContractError(msg)
    files = tuple(_require_str(f, "files[]") for f in files_raw)

    tags_raw = raw.get("tags")
    if not isinstance(tags_raw, list):
        msg = "'tags' must be a JSON array"
        raise ContractError(msg)

    tags: list[Tag] = []
    for i, entry in enumerate(tags_raw):
        if not isinstance(entry, dict):
            msg = f"tags[{i}] must be a JSON object"
            raise ContractError(msg)
        kind = _require_str(entry.get("kind"), f"tags[{i}].kind")
        if kind not in TAG_KINDS:
            allowed = sorted(TAG_KINDS)
            msg = f"tags[{i}].kind must be one of {allowed}, got {kind!r}"
            raise ContractError(msg)
        name = _require_str(entry.get("name"), f"tags[{i}].name")
        path = _require_str(entry.get("path"), f"tags[{i}].path")
        line = _require_int(entry.get("line"), f"tags[{i}].line")
        tags.append(Tag(kind=kind, name=name, path=path, line=line))

    return Request(files=files, tags=tuple(tags))


def compute_symbols(req: Request) -> list[Symbol]:
    """Count internal and external fan-in for every in-scope definition."""
    scope = frozenset(req.files)

    # Keys, not a set of Symbol: dedupes an exact-duplicate def without
    # losing two distinct definitions that merely share a name.
    defs: dict[tuple[str, str, int], None] = {}
    for tag in req.tags:
        if tag.kind == "def" and tag.path in scope:
            defs[(tag.path, tag.name, tag.line)] = None

    calls_by_name: dict[str, list[str]] = {}
    for tag in req.tags:
        if tag.kind == "call":
            calls_by_name.setdefault(tag.name, []).append(tag.path)

    symbols: list[Symbol] = []
    for path, name, line in defs:
        call_paths = calls_by_name.get(name, [])
        internal = sum(1 for p in call_paths if p == path)
        external = sum(1 for p in call_paths if p != path)
        symbols.append(
            Symbol(
                name=name,
                path=path,
                line=line,
                internal=internal,
                external=external,
            ),
        )

    # Deliberate re-sort, never a reliance on the dict's insertion order:
    # (path, name) per the contract, with line as a deterministic
    # tiebreak for two definitions that share both.
    symbols.sort(key=lambda s: (s.path, s.name, s.line))
    return symbols


def render(symbols: list[Symbol]) -> str:
    """Serialise to the exact byte shape the Nim side must also produce."""
    payload = {
        "symbols": [
            {
                "name": s.name,
                "path": s.path,
                "line": s.line,
                "internal": s.internal,
                "external": s.external,
            }
            for s in symbols
        ],
    }
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=True)


def main() -> int:
    """Read stdin, validate, compute, emit once. Never partial output."""
    raw_bytes = sys.stdin.buffer.read()
    try:
        raw_text = raw_bytes.decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        sys.stderr.write(f"fanin: stdin is not valid UTF-8: {exc}\n")
        return 1

    try:
        parsed = json.loads(raw_text)
    except json.JSONDecodeError as exc:
        sys.stderr.write(f"fanin: invalid JSON on stdin: {exc}\n")
        return 1

    try:
        request = parse_request(parsed)
    except ContractError as exc:
        sys.stderr.write(f"fanin: {exc}\n")
        return 1

    symbols = compute_symbols(request)
    sys.stdout.write(render(symbols))
    return 0


if __name__ == "__main__":
    sys.exit(main())
